src/Language/Haskell/TH/Lift/Generics.hs

```diff
@@ -113,8 +194,76 @@ import GHC.Exts (Char(..))
 -- import Language.Haskell.TH.Syntax
 --
 -- foo :: Foo
--- foo = $(lift (Foo 1 'a' "baz"))
+-- foo = $(lift (Foo 1 \'a\' \"baz\"))
 -- @
+
+-- | Generically produce an implementation of 'lift', given a user-provided
+-- (but correct!) package name. The provided package name is ignored for
+-- GHC 7.4 (@base@ 4.5) and later.
+--
+-- === Note
+--
+-- A @'Data.Typeable.Typeable' a@ instance is required for 7.4 <= GHC < 8.0 (4.5 <= @base@ < 4.9).
+#if MIN_VERSION_base (4,9,0) || !MIN_VERSION_base (4,5,0)
+genericLiftWithPkgFallback :: (Quote m, Generic a, GLift (Rep a)) => String -> a -> m Exp
+#else
+genericLiftWithPkgFallback :: (Quote m, Generic a, GLift (Rep a), Typeable a) => String -> a -> m Exp
+#endif
+#if MIN_VERSION_base(4,5,0)
+genericLiftWithPkgFallback _pkg = genericLift
+#else
+genericLiftWithPkgFallback = genericLiftWithPkg
+#endif
+
+#if MIN_VERSION_template_haskell(2,9,0)
+-- | Like 'genericLiftWithPkgFallback', but returns a 'Code' instead of an 'Exp'.
+#if MIN_VERSION_base (4,9,0) || !MIN_VERSION_base (4,5,0)
```

Comment:
Make sure to indent this `#if` one space, as well as other CPP pragmas that follow.